### PR TITLE
feat(speechlm2): add optional MTP training modes

### PR DESCRIPTION
**Pull request by Slyne**
## What

- add opt-in MTP training to `SALMAutomodel`
- support checkpoints with native MTP heads and base LLM checkpoints without MTP by constructing a fresh attached head
- add explicit `joint` and `head_only` training modes, plus a disabled path that suppresses native checkpoint MTP construction
- use Automodel's per-depth MTP-loss API and log per-depth losses
- compute verifier-prefix validation agreement metrics
- preserve packed-sequence/THD masking semantics and partial distributed-checkpoint loading
- bound MTP loss memory with fused linear cross-entropy when `cut_cross_entropy` is installed, with a supported unfused fallback
- keep reusable target, metric, loss-selection, and eval-gating helpers in `speechlm2/parts/mtp.py`

## Why

SpeechLM2 recipes need one consistent MTP path across Lightning-backbone and Nemotron Nano-family LLMs, including base checkpoints that do not already contain MTP weights. The feature remains opt-in so existing recipes are unchanged.

## Automodel dependencies and integration boundary

This PR targets a future Speech Automodel pin that contains:

- [#2802](https://github.com/NVIDIA-NeMo/Automodel/pull/2802): native `compute_mtp_in_eval` support
- [#3414](https://github.com/NVIDIA-NeMo/Automodel/pull/3414): gradient-safe sharded LM-weight materialization and related VLM CP fixes
- [#3525](https://github.com/NVIDIA-NeMo/Automodel/pull/3525): optional per-depth return from `calculate_mtp_loss` (merged)
- [#3543](https://github.com/NVIDIA-NeMo/Automodel/pull/3543): caller-precomputed MTP embeddings, position IDs, and targets for context parallelism

The final Speech dependency pin must be updated after #3543 merges. Until then this PR depends on merged #3525 plus open #3543 and is not compatible with the current `ee0b2261` pin.

Automodel owns native/attached MTP modules and generic MTP loss computation. This PR owns SpeechLM2 configuration, training policy, packed-audio masking, checkpoint/export integration, and validation metrics. The eval helper uses #2802's native model gate while keeping child modules in eval mode.

MTP now supports `cp_size>1` for packed THD with Transformer Engine when `tp_size=1`. Speech prepares future-token embeddings, position IDs, and loss targets in global packed order, masks packed-document boundaries, then applies the same interleaved CP partition as the base inputs before passing the rank-local tensors through #3543's interfaces. BSHD+CP remains unsupported, and MTP still requires `tp_size=1` because the MTP head has no tensor-parallel plan. Teacher-forced verifier-prefix metrics remain disabled under CP because rank-local verifier predictions cannot be shifted across CP boundaries without additional communication.

### Local

- `59 passed`: focused Speech MTP and packed-sequence tests against a combined Automodel #3525 + #3543 checkout
- `68 passed`: broader pretrained, CP-helper, parallel-strategy, and SALM regression coverage; 8 additional SALM fixture cases were not runnable locally because the sandbox could not acquire the existing Hugging Face cache lock
- Black and isort checks pass on all changed Python files
- `git diff --check` passes
- The existing GPU smoke matrix below predates the CP integration and covers `cp_size=1`; Automodel #3543 carries the distributed packed-THD MTP parity validation for the shared CP contract

### IAD / 8×A100 smoke matrix

All final runs used the updated image `nemo-speech-pr16021-cu12-a100.sqsh`, Automodel #3525 head `2bebedc7` (which includes merged #2802), one real-audio optimizer step, one validation batch, and DCP checkpoint save.

| Job | Coverage | Result |
| --- | --- | --- |
| `11830644` | Lightning checkpoint, native MTP head, joint training | passed; `loss=6.000`, `mtp_loss=0.711`, `val_acc=0.16895` |
| `11830645` | Lightning checkpoint, MTP disabled | passed; `loss=6.000`, `val_acc=0.16797`; no MTP loss |
| `11829786` | Nano checkpoint without native MTP, fresh head, joint training | passed; `loss=5.720`, `mtp_loss=1.330`, `val_acc=0.24219` |
| `11829788` | Nano checkpoint without native MTP, strict head-only training | passed; 37.9M trainable parameters; `loss=5.720`, `mtp_loss=1.310`, `val_acc=0.24023` |
| `11829787` | Nano checkpoint without native MTP, MTP disabled | passed; `loss=5.910`, `val_acc=0.24414`; no MTP loss |
| `11829926` | packed THD, fresh attention+MoE head, joint training | passed; `loss=6.410`, `mtp_loss=0.170`, `val_acc=0.23535` |
| `11829925` | packed THD, fresh attention+MoE head, strict head-only training | passed; 37.9M trainable parameters; `loss=6.440`, `mtp_loss=0.169`, `val_acc=0.23730` |
| `11829923` | packed THD, MTP disabled | passed; `loss=6.440`, `val_acc=0.23730`; no MTP loss |

The updated IAD image still does not contain optional `cut_cross_entropy`; MTP-enabled jobs therefore also exercised the supported unfused loss fallback.

### NRT / 8×H100

Job `6038792` passed with the updated image `nemo-speech-pr16021-cu12-h100plus-v3.sqsh`, Automodel #3525 head `2bebedc7`, one real-audio joint-MTP optimizer step, one validation batch, and DCP checkpoint save. The run used the staged local Lightning snapshot and Canary `.nemo` artifact because NRT is offline; the real MCV 2K manifest was used as the bounded training input.

- training: `loss=6.000`, `mtp_loss=0.711`
- validation: `val_acc=0.16602`
- checkpoint: `step=1.ckpt` saved successfully
- cold-cache first step: 326 seconds due to H100 TorchInductor compilation

The image still lacks optional `cut_cross_entropy`, so this run also exercised the supported unfused MTP-loss fallback.

### NRT context-parallel / 8×H100

Job `6089571` passed against Speech head `7dbca42aa0` plus a clean Automodel #3525 + #3543 integration (`bc423748`). The packed-THD run used CP2/EP4/TP1/DP4, Transformer Engine attention/linear layers, one real MCV 2K joint-MTP optimizer step, one validation batch, and distributed checkpointing.

- training: `loss=5.840`, `mtp_loss=1.580` (`train_step_timing=259.0s`)
- validation: `val_acc=0.21289`; teacher-forced verifier-prefix MTP metrics were intentionally skipped under CP
- checkpoints: both `step=1.ckpt` and `step=1-last.ckpt` contain all eight DCP shards plus metadata
- SLURM: completed `0:0` in 9m50s on one 8×H100 node
- environment caveat: NRT's default `TORCH_NCCL_USE_COMM_NONBLOCKING=1` makes the first Mamba CP `all_to_all_single` return `NCCL Error 7: NCCL operation in progress`; the passing control set it to `0`. The container also required the workspace's compatible Lhotse overlay because its bundled Lhotse predates `ais_force_individual`.

### NRT context-parallel / 16×H100

Job `6091814` passed against Speech head `f6345345e8` plus the same clean Automodel #3525 + #3543 integration (`bc423748`). The two-node packed-THD run used CP2/EP8/TP1/inferred-DP8, Transformer Engine attention/linear layers, one finite MCV 2K joint-MTP optimizer step, one validation batch, and distributed checkpointing.

- training: `loss=5.840`, `mtp_loss=1.400` (`train_step_timing=72.60s`)
- validation: `val_acc=0.22070`; the new one-time warning explicitly reported that teacher-forced verifier-prefix MTP metrics are disabled under CP
- checkpoints: both `step=1.ckpt` and `step=1-last.ckpt` contain all sixteen DCP shards plus `.metadata` and `meta.pt`
- SLURM: completed `0:0` in 5m51s on two 8×H100 nodes
- environment: retained `TORCH_NCCL_USE_COMM_NONBLOCKING=0` and the compatible Lhotse overlay from the passing CP2/EP4 control

An initial allocation (`6091503`) stopped before its first batch on a permission error in the cluster config's default indexed-data mirror. The passing retry changed only the input pipeline to the finite, non-indexed map-style MCV setup already validated by job `6089571`; model sources and CP2/EP8 topology were unchanged.

This supersedes #15791 with a clean implementation based on current Speech `main`; unrelated vLLM inference remains follow-up work.

### CODE REVIEW

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 1]
```diff
@@ -1072,3 +1315,271 @@ def oomptimizer_schema(self) -> dict:
...
+    scaling_factor: float = 0.1,
+    num_label_tokens: torch.Tensor | None = None,
+    grad_reduce_group: dist.ProcessGroup | None = None,
+    ignore_index: int = -100,
+    cu_seqlens: torch.Tensor | None = None,
+    seq_idx: torch.Tensor | None = None,
+) -> tuple[torch.Tensor, list[torch.Tensor]]:
+    """Compute aggregate MTP loss and per-head losses for logging.
+
+    This mirrors Automodel's ``calculate_mtp_loss`` but keeps the intermediate
```
Comment (Slyne) [inline nemo/collections/speechlm2/models/salm_automodel.py]: Maybe instead of copying the codes from Automodel, we can file another PR to Automodel to let it return the intermediate loss so we can remove the duplicate codes here.

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 2]
```diff
@@ -374,6 +400,19 @@ def _validate_parallelism_compatibility(self) -> None:
             nvte_fused_attn=nvte_fused_attn,
             device_capability=device_capability,
         )
+        mtp_cfg = self.cfg.get("mtp", None)
+        mtp_enabled = mtp_cfg is not None and bool(mtp_cfg.get("enabled", False))
+        if cp_size > 1 and mtp_enabled:
+            raise ValueError(
+                "SALMAutomodel MTP currently requires cp_size=1. Context parallelism uses an interleaved "
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py]: This is going to be a blocker - can you figure it out @Slyne ?

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 3]
```diff
@@ -517,10 +592,45 @@ def on_validation_epoch_end(self) -> None:
             if lss_vals:
                 self.log("val_lss", torch.stack(lss_vals).mean(), on_epoch=True, sync_dist=True)
 
+        # Multi-Token Prediction teacher-forced agreement metrics. Each accumulated per-head
+        # count is already a prefix count: depth k contributes only when every draft through
+        # depth k agrees with verifier logits from the ground-truth-conditioned validation
+        # forward. This is a cheap quality proxy, not speculative-decoding acceptance: exact
+        # acceptance requires verifier forwards conditioned on each proposed draft prefix.
+        if self._partial_val_mtp_correct:
+            agreement_lengths = []
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py:636]: can we extract this into a method/function, sth like `compute_mtp_agreement_lengths`?

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 4]
```diff
@@ -533,11 +643,15 @@ def validation_step(self, batch: dict, batch_idx: int):
             if dataset_batch is None:
                 continue  # some dataset is exhausted
             inputs = self.prepare_inputs(dataset_batch)
-            forward_outputs = self(
-                inputs["input_embeds"],
-                attention_mask=inputs["attention_mask"],
-                **inputs.get("llm_kwargs", {}),
-            )
+            # Enable MTP only around the validation forward. New Automodel revisions expose
+            # ``compute_mtp_in_eval``; the compatibility path for the currently pinned revision
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py]: Is it possible to import their code directly here?

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 5]
```diff
@@ -1072,3 +1315,271 @@ def oomptimizer_schema(self) -> dict:
                 {"name": "loss_mask", "type": NeuralType(("B", "T"), MaskType()), "seq_length": "output"},
             ],
         }
+
+
+def _build_mtp_loss_fn() -> torch.nn.Module:
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py]: Can we move as much of MTP code as possible into `nemo/collections/speechlm2/parts/mtp.py` ? Just to keep this file's size more contained, but maybe this will be useful in future model classes too.

File: nemo/collections/speechlm2/models/salm_automodel.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/models/salm_automodel.py]: ++

File: nemo/collections/speechlm2/models/salm_automodel.py (hunk 2 above)
Comment (Slyne) [reply nemo/collections/speechlm2/models/salm_automodel.py]: Hey Piotr, in the default automodel MTP loss, it has the labels (not preprocessed for MTP heads) as one of the input tensors in calculate_mtp_loss func and under CP, it is a rank-local TE shard. So rolling it for calculating MTP loss will lead to incorrect loss on the CP shard and packed-sequence boundaries.  We need to preprocess the global labels correctly under CP before we send the preprocessed MTP labels to the MTP loss func. 
This is an accuracy issue for automodel MTP training pipeline with CP.  I plan to raise this issue and fix it later (already have a draft PR, verifying the accuracy now).    
For now, can we disallow the cp training together with MTP in SALM in current PR?

Comment (pzelasko) [reply nemo/collections/speechlm2/models/salm_automodel.py]: OK - for long context we'll need to train MTP as a separate stage with short context, hope that works.

**Review by pzelasko — APPROVED** [review]
